pallets/randomness/src/mock.rs

```diff
@@ -142,6 +142,8 @@ impl Config for Test {
 	type MaxBlockDelay = MaxBlockDelay;
 	type BlockExpirationDelay = MaxBlockDelay;
 	type EpochExpirationDelay = MaxBlockDelay;
+	#[cfg(feature = "runtime-benchmarks")]
+	type KeySetter = ();
```

Comment:
I think we can implement the keys setter as part of the `KeysLookup` trait. A new function of the trait only implemented if runtime-benchmarks is present should do the trick. No need to add a new config type I think

Reply:
`KeysLookup<Id, Keys>` does not have the same generics as `BenchmarkSetKeys<Id, Account, Keys>`. So should we add the the `Account` generic to KeysLookup if it won't use it normally. 

I don't think we can add the generic `Account` only if `#[cfg(feature = "runtime-benchmarks")]`.

Reply:
> should we add the the Account generic to KeysLookup if it would only be used in #[cfg(feature = "runtime-benchmarks")]

actually struggling with this (see CI error), not sure when it'll be working, I'm going to focus on other stuff now

Reply:
This is done now, I agree with the suggestion now that it is done.

Reply:
Something you can do is:

In primitives:
```
/// A Trait to lookup keys from AuthorIds
pub trait KeysLookup<AuthorId, Keys> {
	#[cfg(feature = "runtime-benchmarks")]
	type Account;

	fn lookup_keys(author: &AuthorId) -> Option<Keys>;
	
	#[cfg(feature = "runtime-benchmarks")]
	fn set_keys(id: AuthorId, account: Self::Account, keys: Keys);
} 
```

In pallet-author-mapping:

```
impl<T: Config> KeysLookup<NimbusId, T::Keys> for Pallet<T> {
		#[cfg(feature = "runtime-benchmarks")]
		type Account = T::AccountId;

		fn lookup_keys(author: &NimbusId) -> Option<T::Keys> {
			Self::keys_of(author)
		}
		
		#[cfg(feature = "runtime-benchmarks")]
		/// Sets keys WITHOUT reserving deposit, for benchmarking purposes only
		fn set_keys(id: NimbusId, account: Self::Account, keys: T::Keys) {
			use sp_runtime::traits::Zero;
			MappingWithDeposit::<T>::insert(
				id,
				RegistrationInfo {
					account,
					deposit: Zero::zero(),
					keys,
				},
			);
		}
	}
```
And lastly, in your randomness benchmark, you can add a where clause of the following form:

```
benchmarks! {
	where_clause { where <T::VrfKeyLookup as KeysLookup<NimbusId, VrfId>>::Account: From<T::AccountId> }
```

Reply:
Also dont forget that runtime-benchmarks feature should not be enabled by default. You should add it only when compiled with the flag, i.e.:

In pallet-author-mapping:

```rust
session-keys-primitives = { path = "../../primitives/session-keys", default-features = false}
runtime-benchmarks = [
	"frame-benchmarking",
	"session-keys-primitives/runtime-benchmarks"
]
```

and in randomness:

```rust
session-keys-primitives = { path = "../../primitives/session-keys", default-features = false }
runtime-benchmarks = [
	"frame-benchmarking",
	"pallet-author-mapping/runtime-benchmarks",
	"session-keys-primitives/runtime-benchmarks"
]
```